### ISSUE TITLE
Super H ADD #imm function fault

### ISSUE THREAD

**Issue by its5h1ll** [NationalSecurityAgency/ghidra#5274]
**Describe the bug**
A clear and concise description of the bug.

The ADD #imm function seems to not work, the add institution is not adding the data stored in the referenced address to the selected register. this is making a lot of the code incomplete/incorrect. Is there a way to edit the ghidra code to change the way the add #imm code works? 

To make the code make perfect sense I have to patch the instruction and use the data sorted in the referenced location as a normal add instruction, and this is taking far too much time. When opening the patch instruction it's clear to see there's some bug because the data shown in no way relates to anything in the stored address or anything in the local area. What can I do to correct this
![IMG_20230429_135840](https://user-images.githubusercontent.com/132140415/235301387-206abbcc-b722-45fb-bf0e-d8e9e642c0d4.jpg)
![IMG_20230429_135848](https://user-images.githubusercontent.com/132140415/235301391-513845ac-6bab-4a4c-81fa-3734eaa84001.jpg)
![IMG_20230429_135840](https://user-images.githubusercontent.com/132140415/235301421-21ef33c1-4452-47d5-9ddd-557abff192c9.jpg)
![IMG_20230429_135848](https://user-images.githubusercontent.com/132140415/235301423-341e6f45-40ff-46e9-b799-6ccd3dde5cea.jpg)

**To Reproduce**

**Expected behavior**

**Screenshots**

**Attachments**
If applicable, please attach any files that caused problems or log files generated by the software.

**Environment (please complete the following information):**
 - OS: [e.g. macOS 10.14.2]
 - Java Version: [e.g. 11.0]
 - Ghidra Version: [e.g. 10.1.4]
 - Ghidra Origin: [e.g. official GitHub distro, third party distro, locally built] 

**Additional context**

**Comment by fenugrec** [comment]
You know you can copy-paste text instead of *taking a photo* of the screen ? Or at the very least, proper screenshots ? Also, those 4 photos are nearly useless; they all show the same area, while still hiding the important information : the raw opcode bytes, left columns.

Have mercy, edit and clean up your post.

**Comment by pjsoberoi** [comment]
> The ADD #imm function seems to not work, the add institution is not adding the data stored in the referenced address to the selected register.

Per the docs the ADD #imm instruction is not supposed to reference memory. 

http://shared-ptr.com/sh_insns.html
```
void ADDI (int i, int n)
{
  if ((i & 0x80) == 0)
    R[n] += (0x000000FF & (long)i);
  else
    R[n] += (0xFFFFFF00 | (long)i);

  PC += 2;
}
```

SuperH.sinc
```
:add    simm_00_07,rn_08_11  is opcode_12_15=0b0111 & rn_08_11 & simm_00_07
{
    rn_08_11 = rn_08_11 + simm_00_07;
}
```

SuperH4.sinc
```
I_0t:    "#"^I_0  is I_0 { tmp:4 = I_0; export tmp; }

# Binary Addition
# pattern 0111nnnniiiiiiii
# text    add #<imm>,<REG_N>
# arch    arch_sh_up
:add I_0t,N_0t              is OP_0=0x7 & N_0t & I_0t {

	N_0t = I_0t + N_0t; # NOTE I_0t already signed extended
}
```

Neither SuperH or SuperH reference memory. SuperH4 does seem to have a constructor that does something with the bitness, but I'm not sure if this is necessary. It also is not referencing memory. 

To me it looks like everything is working as intended. Please let me know if I'm missing something. Thanks.

**Comment by GhidorahRex** [comment]
> Neither SuperH or SuperH reference memory. SuperH4 does seem to have a constructor that does something with the bitness, but I'm not sure if this is necessary. It also is not referencing memory.
> 
> To me it looks like everything is working as intended. Please let me know if I'm missing something. Thanks.

Neither should be referencing memory, and the pcode between them is identical. I couldn't find any bitness stuff going on with the superH4 add immediate other than the statement that it didn't need to be sign extended. I'm still not entirely sure what the issue is here either? Is it referencing memory when it shouldn't? Does the user think it should be referencing memory?

**Comment by fenugrec** [comment]
> In the above screenshots it does not appear that negative values are being sign-extended.

Ah so that's what we're supposed to see. Indeed on "all" SuperH (I checked on -2E, -2A, -3, -4AL) the `ADD #imm8 ` form should always sign-extend.

**Comment by GhidorahRex** [comment]
> @GhidorahRex,
> 
> Reviewing the attached screenshots, I believe the sign extension is the issue. Per the documentation:
> 
> > The 8-bit immediate value is sign-extended to 32 bits, which allows it to be used for immediate subtraction or decrement operations.
> 
> In the above screenshots it does not appear that negative values are being sign-extended.
> 
> SuperH4 has this constructor:
> 
> `I_0t: "#"^I_0 is I_0 { tmp:4 = I_0; export tmp; }`
> 
> I can submit a PR tonight for SuperH.

The actual value of the computation *should* be sign extended. I don't have a superH binary to test with right now, and the screenshots above do not have the pcode field enabled. In that case, it's just an issue with the display, but it seems weird.

**Comment by fenugrec** [comment]
> The actual value of the computation _should_ be sign extended. I don't have a superH binary to test with right now, and the screenshots above do not have the pcode field enabled.

Here's one bin I've posted before  : [6Z68A.zip](https://github.com/NationalSecurityAgency/ghidra/files/8996477/6Z68A.zip)
```
        0000010a 7f f8           add        -0x8,r15
                                                      r15 = INT_ADD r15, 0xfffffff8:4
```
Not sure how OP got it to show an xref to "DAT_0000FFF8"

**Comment by fenugrec** [comment]
Maybe some of this (the xref part) is vaguley related to #4387

**Comment by GhidorahRex** [comment]
Your disassembly looks like what I would expect it to look like, and the sleigh matches my expectation of how this instruction should behave.

**Comment by fenugrec** [comment]
> Your disassembly looks like what I would expect it to look like, and the sleigh matches my expectation of how this instruction should behave.

Any reason why SH and SH4 show up slightly differently in the pcode representation ?

**Comment by pjsoberoi** [comment]
> Any reason why SH and SH4 show up slightly differently in the pcode representation ?

SH4 has the constructor:

`I_0t: "#"^I_0 is I_0 { tmp:4 = I_0; export tmp; }`

From your disassembly sign extensions appear to be happening on SH with the need of the constructor. @GhidorahRex can you help explain what the constructor is doing. Thanks.

**Comment by GhidorahRex** [comment]
That sub-constructor isn't doing anything other than fixing the size it to 32-bits. Theoretically there could be a reason for it - if it always needed to be 32-bits regardless of register assignment or something, for example. But I think it was just to make the formatting of the sleigh code look more compact.

The development of the two modules were entirely unrelated. The 4H was internally developed, and the SuperH was a github pull request. So there's not a lot of similarity between the two modules, even though they're as closely related as 68k and Coldfire or x86 and x86-64.

**Comment by its5h1ll** [comment]
> > The ADD #imm function seems to not work, the add institution is not adding the data stored in the referenced address to the selected register.
> 
> Per the docs the ADD #imm instruction is not supposed to reference memory.
> 
> http://shared-ptr.com/sh_insns.html
> 
> ```
> Description
> Adds together the contents of general register Rn and the immediate value and stores the result in Rn. The 8-bit immediate value is sign-extended to 32 bits, which allows it to be used for immediate subtraction or decrement operations.
> 
> Operation
> void ADDI (int i, int n)
> {
>   if ((i & 0x80) == 0)
>     R[n] += (0x000000FF & (long)i);
>   else
>     R[n] += (0xFFFFFF00 | (long)i);
> 
>   PC += 2;
> }
> ```
> 
> SuperH.sinc
> 
> ```
> :add    simm_00_07,rn_08_11  is opcode_12_15=0b0111 & rn_08_11 & simm_00_07
> {
>     rn_08_11 = rn_08_11 + simm_00_07;
> }
> ```
> 
> SuperH4.sinc
> 
> ```
> I_0t:    "#"^I_0  is I_0 { tmp:4 = I_0; export tmp; }
> 
> # Binary Addition
> # pattern 0111nnnniiiiiiii
> # text    add #<imm>,<REG_N>
> # arch    arch_sh_up
> :add I_0t,N_0t              is OP_0=0x7 & N_0t & I_0t {
> 
> 	N_0t = I_0t + N_0t; # NOTE I_0t already signed extended
> }
> ```
> 
> Neither SuperH nor SuperH4 reference memory. SuperH4 does seem to have a constructor that does something with the bitness, but I'm not sure if this is necessary. It also is not referencing memory.
> 
> To me it looks like everything is working as intended. Please let me know if I'm missing something. Thanks.

As per the software manual, ADD #imm function is immediate addressing, so in my understanding it takes the data stored in the mentioned address and adds it to the specified register. When this is entered manually this makes the code much more sensible and fits perfectly with the code, example when manually inputting the data in the referenced address as a normal ADD function it points to an address which is in the peripherals and it is a specified address of an input of the chip, when it's left as is it points to something that is not referenced in the hardware manual at all. It makes perfect sense because in a series of functions the ADD #imm function should point to an address where 01 is stored and this should be added to the PR, however the bug tried to add -4 to the PR which is clearly not right

**Comment by its5h1ll** [comment]
> > Neither SuperH or SuperH reference memory. SuperH4 does seem to have a constructor that does something with the bitness, but I'm not sure if this is necessary. It also is not referencing memory.
> > To me it looks like everything is working as intended. Please let me know if I'm missing something. Thanks.
> 
> Neither should be referencing memory, and the pcode between them is identical. I couldn't find any bitness stuff going on with the superH4 add immediate other than the statement that it didn't need to be sign extended. I'm still not entirely sure what the issue is here either? Is it referencing memory when it shouldn't? Does the user think it should be referencing memory?

It should definitely be referring a memory address, it should use the address as the data for the add function. "ADD #imm,Rn Rn + imm → Rn" or am I confused 😅 I'm not a programmer or software engineer so I won't say I'm certain but when edited to a normal ADD function adding the data stored in the mentioned address is makes perfect sense. Example one such code is "ADD #immm 0000FFFA to PR" and stored in 0000FFFA is 00000001, alternatively when not manually edited it tries to add -04 which is surely not correct?

**Comment by its5h1ll** [comment]
Please see the software manual screenshot, I have highlighted the specific part which says the direct memory address is used for the ADD function. When I look at the data in the referenced address is 32bit FLOAT Hi-Lo it makes perfect sense to add this immediate data. Is there some way I can edit Ghidra to make this disassembled correctly because changing everything manually means I have to clear the disassembled code and disassemble it again after for it to show correctly, and then the code further along is not correct....I was stuck at one point and couldn't understand how data in a ram address was moved to another ram address even though I knew it must be moved from the peripheral to the ram and then moved into ram again after some arithmetic changing it from A/D counts into a voltage value, but the voltage value was not moved to another ram address after which I knew it must do. Having changed a few sequences of this incorrect ADD #imm function manually it now has the function where the ram address has the data added to it which makes it point to that voltage and then that voltage is moved into a sensor linearization calculation function to output the sensor value.

**Comment by its5h1ll** [comment]
@GhidorahRex 

I will post some screenshots later of the code for better understanding, but there are multiple instances of ADD #imm function which reference DAT_ 0X addresses, and in those addresses when I view the data which is shown in 32bit float (all rom memory is in 32bit float including pointers for the interrupts, instructions are in 16bit) the whole software irons out and makes more sense

**Comment by its5h1ll** [comment]
@fenugrec 

> You know you can copy-paste text instead of **taking a photo** of your monitor ? Or at the very least, proper screenshots ? Also, those 4 photos are nearly useless; they all show the same area, while still hiding the important information : the raw opcode bytes, left columns.
> 
> Have mercy, edit and clean up your post.

Sorry I'm not a programmer so I don't know the correct format, I'm just winging it/learning as I go.
I'll make sure my next post is done better, with full size screenshots and a copy of the data string

**Comment by GhidorahRex** [comment]
When an assembly instruction mentions immediate data it means a number encoded as part of the instruction. For the superH, the instructions which read from memory have an `@`-sign prefacing the data that has the data address. The code for those instructions looks different as well. For example, take `mov.b @Rm,Rn`:

```
void MOVBL (int m, int n)
{
  R[n] = Read_8 (R[m]);
  if ((R[n] & 0x80) == 0)
    R[n] &= 0x000000FF;
  else
    R[n] |= 0xFFFFFF00;

  PC += 2;
}
```

It accesses memory through the `Read_8(R[m]);` pseudofunction. The `addi` instruction does not have any such reference to memory. This instruction is correct as it is.

**Comment by its5h1ll** [comment]
> When an assembly instruction mentions immediate data it means a number encoded as part of the instruction. For the superH, the instructions which read from memory have an `@`-sign prefacing the data that has the data address. The code for those instructions looks different as well. For example, take `mov.b @Rm,Rn`:
> 
> ```
> void MOVBL (int m, int n)
> {
>   R[n] = Read_8 (R[m]);
>   if ((R[n] & 0x80) == 0)
>     R[n] &= 0x000000FF;
>   else
>     R[n] |= 0xFFFFFF00;
> 
>   PC += 2;
> }
> ```
> 
> It accesses memory through the `Read_8(R[m]);` pseudofunction. The `addi` instruction does not have any such reference to memory. This instruction is correct as it is.

@GhidorahRex 

I have attached some screenshots now for some better explanation. I am at a loss as there are multiple instances of this in the code as you can see in one of the screenshots. 
![1add #DAT_](https://user-images.githubusercontent.com/132140415/236562538-ece24937-04ae-4a9a-9674-0c6b9834f476.jpg)
![2add #DAT_](https://user-images.githubusercontent.com/132140415/236562551-8c185630-2d31-4bf0-a3ed-8c69559fed31.jpg)
![add #DAT_](https://user-images.githubusercontent.com/132140415/236562562-11bc1cca-93cb-4d15-97f9-deeed926b320.jpg)

In some instances if I change the code to add the data sorted in the mentioned memory address it seems to have a more sensible output. 

As writing this I think I have just worked out what the problem is, its saying to add FE (-2 dec) to the named register but its showing as if it's asking to add the data in address FFFE.

So it is definitely not in any case supposed to add the data in the named address?

Thanks for your help so far

**Comment by fenugrec** [comment]
> As per the software manual, ADD #imm function is immediate addressing, so in my understanding it takes the data stored in the mentioned address and adds it to the specified register. 

No. Read that page of the SH* software manual again;

```
ADDI(long i,long n) /* ADD #imm,Rn */ {
	if ((i&0x80)==0)
		R[n]+=(0x000000FF & (long)i);
	else
		R[n]+=(0xFFFFFF00 | (long)i);
	PC+=2;
}
```

As Ghidorarex explained,

> the instructions which read from memory have an @-sign prefacing the data that has the data address

Compare the previous description of ADDI with, for example, MOVLL

```
MOVLL(long m,long n) /* MOV.L @Rm,Rn */ {
	R[n]=Read_Long(R[m]);
	PC+=2;
}
```

The `add #imm8,Rn` form modifies only Rn. It never reads nor writes from memory, ever.

**Comment by its5h1ll** [comment]
> > As per the software manual, ADD #imm function is immediate addressing, so in my understanding it takes the data stored in the mentioned address and adds it to the specified register.
> 
> No. Read that page of the SH* software manual again;
> 
> ```
> ADDI(long i,long n) /* ADD #imm,Rn */ {
> 	if ((i&0x80)==0)
> 		R[n]+=(0x000000FF & (long)i);
> 	else
> 		R[n]+=(0xFFFFFF00 | (long)i);
> 	PC+=2;
> }
> ```
> 
> As Ghidorarex explained,
> 
> > the instructions which read from memory have an @-sign prefacing the data that has the data address
> 
> Compare the previous description of ADDI with, for example, MOVLL
> 
> ```
> MOVLL(long m,long n) /* MOV.L @Rm,Rn */ {
> 	R[n]=Read_Long(R[m]);
> 	PC+=2;
> }
> ```
> 
> The `add #imm8,Rn` form modifies only Rn. It never reads nor writes from memory, ever.

@fenugrec 

I had a bit of confirmation bias when reading the software manual earlier, like I say I'm not a programmer so the whole section here....
Operation:
ADD(long m,long n) /* ADD Rm,Rn */
{
R[n]+=R[m];
PC+=2;
}
ADDI(long i,long n) /* ADD #imm,Rn */
{
if ((i&0x80)==0) R[n]+=(0x000000FF & (long)i);
else R[n]+=(0xFFFFFF00 | (long)i);
PC+=2;

is basically alien language to me. I have no technical understanding on how to read that language, I just read the description of the  and that has been enough to get me by up until this point. The thing that threw me off was the add #DAT_ because it didn't click that the mistake was the #DAT_ was not supposed to be there, I thought the problem was that it was not using the data in the #DAT_ address as the information to add.

I'll have to go back and change some of the code i changed earlier which is rather annoying because it got me out of a dead end i was stuck in, and it made a lot of things make more sense which is quite confusing as to why that happened. The sensor data seemed to move exactly where it should but again that could be confirmation bias or it may just have been coincidence that it ended up in a sensible place

**Comment by its5h1ll** [comment]
So the question still remains, how do I make the add #DAT_ show correctly as add #-0xn? It's making the code hard to follow

**Comment by pjsoberoi** [comment]
@its5h1ll : Just curious, what version of Ghidra are you using?

**Comment by pjsoberoi** [comment]
@its5h1ll : Are you able to share the binary? In the screenshot you linked above, the add #imm instruction is preceded by bf/s which makes use of a delay slot. Unlikely, but I want to rule out issues with the delay slot.

**Comment by its5h1ll** [comment]
> @its5h1ll : Just curious, what version of Ghidra are you using?

@pjsoberoi 

In the About Ghidra tab it says version 10.1.3

I don't think its anything related to the delay in instructions when jumping and branching etc the add #DAT_ instruction is not just limited to branches it's literally every instruction that does an immediate subtraction (adding a negative number)

**Comment by GhidorahRex** [comment]
@its5h1ll Your clarification and the new screenshots clear it up. Your issue is that instructions referencing an immediate value are creating references to data locations based on the immediate value. e.g. `-0x14` references `#DAT_0000ffec`, correct?

**Comment by its5h1ll** [comment]
> @its5h1ll Your clarification and the new screenshots clear it up. Your issue is that instructions referencing an immediate value are creating references to data locations based on the immediate value. e.g. `-0x14` references `#DAT_0000ffec`, correct?

@GhidorahRex 

Yes exactly, it's referencing a data location but from what I've seen, when I select "patch instruction" it seems to know to add the negative number, it's just displaying as #DAT_ for some reason. Also I feel like this is affecting the C code output on the decompiler function, because compared to the C code that was produced from a 68k processor I disassembled as my first project, the C code output of this SuperH is nonsense....or maybe the 68k is just a more intuitive processor language and the C code is more like a maths equation and not proper coding like this SuperH. Would I be correct in saying that?

**Comment by pjsoberoi** [comment]
> In the About Ghidra tab it says version 10.1.3

Do you mind trying in 10.2.3? That is the latest stable.

**Comment by emteere** [comment]
It looks like the DAT_FFEC was just a bad reference created during analysis.  It shouldn't affect the decompiler, but can cause confusion.  The reference isn't actually reading the memory location, it would just be a DATA ptr to it.
___
The correct memory read of the location referred to by register R11 at address C58E is made so the value being tracked in R11 does appear to be working correctly.
The sequence of instructions:
```
    MOV.W  DAT_C598, r10    ; r10 = 0xF6D8
    MOV      r11, R10              ;  r11 = 0xF6D8
    ADD       #-0x14, r11         ; r11 = 0xF6C4
    MOV.B   @r11, r1               ; r11 = the address of TCNT10B
```
There may be some changes you can make to the analysis options to stop this from occuring.
I'm not sure which analyzer is creating the reference.
- You can try turning off the SuperH4ConstRefAnalzyer, then clear the instruction with the reference and then re-disassemble to see if the value comes back.  You really want this analyzer on, but it would be good to know which analyzer is making the bad reference.
- Make sure the ScalarOperandAnalyzer is off.
- You can also try turning off SuperH4ConstRefAnalyzer->StoredValuePointerAnalysis and/or Function parameter/return pointer analysis.

I don't think the reference is coming from the SuperH4ConstRefAnalyzer, but it is possible.
___
Normally loads/stores of a value and passing a value as a parameter can cause a speculative reference.  A load/store from/to a calculated address will cause a read/write reference.  This might be better in 10.2.x or 10.3, but speculative references like this can occur.  Bad references can also occur if something went wrong in the constant value tracing.
___
If you think the generated C-code isn't as good on the SuperH, that can be because of extra memory bank segment offsets or other processor specific idiosyncrasies.  As you add more data types to the code, say to a parameter to a called function, things may improve.

**Comment by fenugrec** [comment]
> when I select "patch instruction"

FYI, I've never had to do that while doing plain reverse engineering. If you're just starting out, I don't think it should be your first reflex when something is strange or unclear.

That said, it's true SuperH has  some issues, e.g. #4387

**Comment by fenugrec** [comment]
this can probably be closed ? The `add #imm8` opcode implementation is correct

**Comment by GhidorahRex** [comment]
@fenugrec agreed. If there is still an issue related to this, please open a new ticket.
